File: image_processing/diff.py
import logging
from PIL import Image, ImageChops
import numpy as np

logger = logging.getLogger("agent-runner")

def generate_diff_image(before_path, after_path, diff_output_path):
    # Process only local files
    img1 = Image.open(before_path).convert('RGB')
    img2 = Image.open(after_path).convert('RGB')

    # Instead of resizing, create new white images with maximum dimensions
    max_width = max(img1.size[0], img2.size[0])
    max_height = max(img1.size[1], img2.size[1])

    # Create white background images
    new_img1 = Image.new('RGB', (max_width, max_height), 'white')
    new_img2 = Image.new('RGB', (max_width, max_height), 'white')

    # Paste original images onto white backgrounds (this will preserve their original sizes)
    new_img1.paste(img1, (0, 0))
    new_img2.paste(img2, (0, 0))

    # Save transformed images.
    new_img1.save(before_path.replace('.png', '-resized.png'))
    new_img2.save(after_path.replace('.png', '-resized.png'))

    # Create a diff image
    diff = ImageChops.difference(new_img1, new_img2)

    # Create a mask of non-zero differences
    diff_np = np.array(diff)
    mask = (diff_np != 0).any(axis=2)

    # Convert mask to RGBA format for transparency
    result_np = np.zeros((max_height, max_width, 4), dtype=np.uint8)

    # Set red color with full opacity where differences exist
    result_np[mask, 0] = 255  # R channel
    result_np[mask, 3] = 255  # Alpha channel (fully opaque)

    # Create final image from numpy array
    diff_highlight = Image.fromarray(result_np, 'RGBA')

    # Save the result as PNG to preserve transparency
    diff_highlight.save(diff_output_path, format='PNG')
    logger.info(f"Diff image saved at: {diff_output_path}")

File: src-typescript/diff/diff.ts
import sharp from "sharp";
import pixelmatch from "pixelmatch";
import { PNG } from "pngjs";

/**
 * Generate a diff image comparing two screenshots.
 *
 * It uses 2 screenshots and pads the smallest image with a white background so that they match and can be
 * compared using pixelmatch.
 *
 * @param beforePath
 *  Path to the before screenshot
 * @param afterPath
 *  Path to the after screenshot
 * @param diffOutputPath
 *  Path where the diff image will be saved
 *
 * @returns void
 *  Writes the diff image to the diffOutputPath
 */
export async function generateDiffImage(
  beforePath: string,
  afterPath: string,
  diffOutputPath: string
): Promise<void> {
  // Load both images and convert to RGB
  const img1 = sharp(beforePath);
  const img2 = sharp(afterPath);

  // Get image dimensions
  const metadata1 = await img1.metadata();
  const metadata2 = await img2.metadata();

  if (
    !metadata1.width ||
    !metadata1.height ||
    !metadata2.width ||
    !metadata2.height
  ) {
    throw new Error("Failed to read image dimensions");
  }

  // Calculate maximum dimensions
  const maxWidth = Math.max(metadata1.width, metadata2.width);
  const maxHeight = Math.max(metadata1.height, metadata2.height);

  // Create white background images
  const whiteBackground1 = sharp({
    create: {
      width: maxWidth,
      height: maxHeight,
      channels: 3,
      background: { r: 255, g: 255, b: 255 },
    },
  });

  const whiteBackground2 = sharp({
    create: {
      width: maxWidth,
      height: maxHeight,
      channels: 3,
      background: { r: 255, g: 255, b: 255 },
    },
  });

  // Convert original images to RGB and get buffers
  const rgbImg1Buffer = await img1.removeAlpha().toBuffer();
  const rgbImg2Buffer = await img2.removeAlpha().toBuffer();

  // Composite original images onto white backgrounds
  const newImg1 = whiteBackground1.composite([
    { input: rgbImg1Buffer, top: 0, left: 0 },
  ]);

  const newImg2 = whiteBackground2.composite([
    { input: rgbImg2Buffer, top: 0, left: 0 },
  ]);

  // Save transformed images (resized versions)
  const beforeResizedPath = beforePath.replace(".png", "-resized.png");
  const afterResizedPath = afterPath.replace(".png", "-resized.png");
  await newImg1.png().toFile(beforeResizedPath);
  await newImg2.png().toFile(afterResizedPath);

  // Get raw RGBA pixel data for pixelmatch
  // pixelmatch expects RGBA buffers (4 channels: R, G, B, A)
  const img1Buffer = await newImg1
    .ensureAlpha()
    .raw({ depth: "uchar" })
    .toBuffer({ resolveWithObject: false });
  const img2Buffer = await newImg2
    .ensureAlpha()
    .raw({ depth: "uchar" })
    .toBuffer({ resolveWithObject: false });

  // Create output PNG for diff
  const diff = new PNG({ width: maxWidth, height: maxHeight });

  // Use pixelmatch to compute differences
  // pixelmatch outputs RGBA data with red highlights where differences exist
  pixelmatch(img1Buffer, img2Buffer, diff.data, maxWidth, maxHeight, {
    threshold: 0.1,
    includeAA: false,
    alpha: 0.1,
  });

  // Save the diff image using sharp
  await sharp(diff.data, {
    raw: {
      width: maxWidth,
      height: maxHeight,
      channels: 4,
    },
  })
    .png()
    .toFile(diffOutputPath);

  console.log(`Diff image saved at: ${diffOutputPath}`);
}
